client/protocols/rate_limit.py

```diff
@@ -0,0 +1,81 @@
+from client.protocols.common import models
+from client.exceptions import ProtocolException
+import time
+
+
+BACKOFF_POLICY = (0, 0, 180, 300) # 1-2 no penalty, penalty in seconds after
+
+
+##################
+# Secure Storage #  TODO: refactor persistance to one centralized location
+##################
+
+import shelve # persistance
+import hmac # security
+
+SHELVE_FILE = '.ok_storage'
+SECURITY_KEY = 'uMWm4sviPK3LyPzgWYFn'.encode('utf-8')
+
+def mac(value):
+    mac = hmac.new(SECURITY_KEY)
+    mac.update(repr(value).encode('utf-8'))
+    return mac.hexdigest()
+
+def check(root, key):
+    key = '{}-{}'.format(root, key)
+    with shelve.open(SHELVE_FILE) as db:
+        return key in db
+
+def store(root, key, value):
+    key = '{}-{}'.format(root, key)
+    with shelve.open(SHELVE_FILE) as db:
+        db[key] = {'value': value, 'mac': mac(value)}
```

Comment:
You're totally right, it's really clunky.  The idea was I wanted to develop this as fast as possible to get it ready for the first assignments, and worry about making clean abstractions later.  This would allow us to thoroughly test that rate-limiting does work as intended before first assignments that use them are released.

I think abstracting should wait until we try and make a refactor shift to have all persistent storage managed via the `.ok_storage` file.  There are other places (like #194) where having this abstraction would be useful, and it'd be great to make this shift at once.  I can make an issue for this the meanwhile?